### ISSUE TITLE
Use of Python round() inbuilt function inside Functional.CenterCrop() NonDeterministic

### ISSUE THREAD

**Issue by revorg7** [pytorch/vision#5274]
### 🐛 Describe the bug

Barring the padding process, CenterCrop is implemented with these lines:
crop_top = int(round((image_height - crop_height) / 2.0))
crop_left = int(round((image_width - crop_width) / 2.0))
return crop(img, crop_top, crop_left, crop_height, crop_width)

Note that round(11/2)=6 and round(9/2)=4, i.e., python round() selects floor() sometimes, and sometimes ceil(). This causes crop_top and crop_left coordinates to be non-deterministic (depending on input sizes). 

Remedy: Just remove the round() since you already have int() outside (change it to floor() to avoid further confusion)

### Versions

Collecting environment information...
PyTorch version: 1.9.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.5 (Ootpa) (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.3.4

Python version: 3.6.8 (default, Sep  9 2021, 07:49:02)  [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)] (64-bit runtime)
Python platform: Linux-4.18.0-348.2.1.el8_5.x86_64-x86_64-with-redhat-8.5-Ootpa
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Quadro RTX 6000
Nvidia driver version: 470.57.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.9.1+cu111
[pip3] torchvision==0.10.1+cu111
[conda] Could not collect

cc @vfdev-5 @datumbox @fmassa @pmeier

**Comment by vfdev-5** [comment]
@revorg7 python round is a deterministic operation as `round(value)` always gives the same output. 
Concerning your point about that round(5.5) -> 6 and round(4.5) -> 4 is related to how floating point works. See https://docs.python.org/3/library/functions.html#round

Your suggestion to remove `round` would introduce a backward incompatibility.

Can you please give more details on what is your exact issue with the center_crop ? Thanks

**Comment by revorg7** [comment]
@vfdev-5 , OK, Let me "correct" my terminology, its a "input dependent deterministic function". In general, "input dependent" functions are not desirable. E.g. Take:

import torch
from torchvision.transforms import CenterCrop

a=torch.Tensor(range(25)).view(5,5)
b=torch.Tensor(range(49)).view(7,7)
c=torch.Tensor(range(81)).view(9,9)

CC= CenterCrop((2,2))

CC(a)
CC(b)
CC(c)

You will notice that the center crops of these images oscillate between being shifted top-left to bottom-right, depending on size of input image. I have attached an image to explain this concept more clearly. 
<img width="572" alt="ExplainationOfCenterCropProblem" src="https://user-images.githubusercontent.com/17990330/151681399-fda39a7c-3fb3-4136-b6f3-768736b05bba.png">

**Comment by vfdev-5** [comment]
> Hence the user never knows what "padding" to expect from CenterCrop.

@revorg7 I understand what you say but I'm not sure if this is something really matters.
For example, I have an image of size 500, 323 that I'd like to center crop to 223, 224. Then, the i'll have the following:
```python
 (image_height - crop_height) / 2.0, (image_width - crop_width) / 2.0
> 49.5, 138.5
```
to determine crop_top, crop_left. Python round will produce: 50, 138. Comparing that with 49, 138 on large image, visually there is no quite difference.
![image](https://user-images.githubusercontent.com/2459423/151840297-49996c54-594f-417b-b636-f1fb616c2b38.png)

I'm trying to understand why `round()` behaviour could be important to the users ?

As for BC-compatibility, we can add an option to customize `fn=lambda x: int(round(x))` and `crop_top = fn((image_height - crop_height) / 2.0))`. This could keep BC and users concerned with python round could use just `lambda x: int(x)` or `lambda x: int(x + 0.5)`.

**Comment by revorg7** [comment]
@vfdev-5 , Sure for high resolution images, it doesn't have much affect, but that may not be the only use case. Consider language models or 2D structured gene sequence as input, one never knows. Regardless, an undefined(hidden) code behaviour is never good. For BC-compatibility, I concur to the idea of using lambda function. 
Finally, one may still be missing the point that function composition int(round(.)) was not needed in the first place, since output of round(.) is already of type int.

**Comment by vfdev-5** [comment]
> One concrete example is when CenterCrop is used in the bottleneck area of a CNN, say AE, which is usually a low resolution 2D filtermap, there its effects would be significant and the network designer would be unaware of it.

For a fixed feature map size the center crop will be the same. On the other hand, if feature map crop uses parameters such that the difference is even, then top-left corner will be simply integer, e.g. round((32 - 8) / 2) = round(12) = 12
On the other hand, I do not understand why if using your numbers like round((7 - 2) / 2) = round(5/2) = round(2.5) can't be 2 and should be 3 ?

> Regardless, an undefined(hidden) code behaviour is never good.

I'd say you have to file an issue to python then :) Python does something like : round(even_base + 0.5) -> even_base, round(odd_base + 0.5) -> odd_base + 1 and it is well defined, IMO. Another option we could do `round(x)` as `int(x + 0.5)` which does basically ceil of integer + 0.5 cases and do not rely on python round.

> Finally, one may still be missing the point that function composition int(round(.)) was not needed in the first place, since output of round(.) is already of type int.

Today transformation support torch jit script and adding `int(round(x))` was necessary for jit.

**Comment by revorg7** [comment]
> For a fixed feature map size the center crop will be the same. On the other hand, if feature map crop uses parameters such that the difference is even, then top-left corner will be simply integer, e.g. round((32 - 8) / 2) = round(12) = 12 On the other hand, I do not understand why if using your numbers like round((7 - 2) / 2) = round(5/2) = round(2.5) can't be 2 and should be 3 ?

You are simply diverting attention to use cases where the issue doesn't arise. Its not rocket science to see that it won't happen for arguments divisible by two.

> I'd say you have to file an issue to python then :) 

I didn't, because it can simply be avoided by not using round() anymore, unless ofcourse you insist on using it anyways.

>Python does something like : round(even_base + 0.5) -> even_base, round(odd_base + 0.5) -> odd_base + 1 and it is well defined, IMO. 

Again, You are simply repeating what we both know, and when I said earlier that round(11/2)=6 and round(9/2)=4. I would call it properly defined, but not "well" defined. Scientifically, for an interval [i,i+1], you either choose mapping [i,i+1) -> i (open interval) or [i,i+1] -> i (closed interval), but not both. That is the reason I don't consider it well defined.

>Another option we could do `round(x)` as `int(x + 0.5)` which does basically ceil of integer + 0.5 cases and do not rely on python round.
So do it. Finally, I don't see any need for further discussion here. I leave it upto you to make(or not) the changes and close the issue.

**Comment by datumbox** [comment]
@revorg7 Thanks for the feedback. 

We had a discussion with the team and unfortuntaely for BC reasons we can't change the `round()` to `floor()/ceil()` (this will have downstream effects on pretrained models and existing pipelines). As discussed above, for the most common cases the 1 pixel difference shouldn't matter much and for those that it is critical (such as the ones you mentioned above), we encourage the users to rely on a custom implementation.
